modelzoo/DBMTL/train.py

```diff
@@ -0,0 +1,900 @@
+            dnn_optimizer = tf.train.AdagradDecayOptimizer(
+                learning_rate=self._learning_rate,
+                global_step=self.global_step)
+        else:
+            raise ValueError("Optimizer type error.")
+
+        train_ops = []
+        update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
+        with tf.control_dependencies(update_ops):
+            train_ops.append(
+                dnn_optimizer.minimize(self.loss,
+                                       var_list=tf.get_collection(
+                                           tf.GraphKeys.TRAINABLE_VARIABLES,
+                                           scope='dnn'),
+                                       global_step=self.global_step))
+        self.train_op = tf.group(*train_ops)
+    
+    def _create_metrics(self):
+        self.acc, self.acc_op = tf.metrics.accuracy(labels=self._label,
+                                                    predictions=self.output)
+        self.auc, self.auc_op = tf.metrics.auc(labels=self._label,
+                                               predictions=self.probability,
+                                               num_thresholds=1000)
+        tf.summary.scalar('eval_acc', self.acc)
+        tf.summary.scalar('eval_auc', self.auc)
+
+
+# generate dataset pipeline
+def build_model_input(filename, batch_size, num_epochs):
+    def parse_csv(value):
+        tf.logging.info('Parsing {}'.format(filename))
+        HASH_defaults = [[" "] for i in range(0, len(HASH_INPUTS))]
+        label_defaults = [[0] for i in range (0, len(LABEL_COLUMNS))]
+        IDENTITY_defaults = [[0] for i in range(0, len(IDENTITY_INPUTS))]
+        column_headers = LABEL_COLUMNS + HASH_INPUTS + IDENTITY_INPUTS
+        record_defaults = label_defaults + HASH_defaults + IDENTITY_defaults
+        columns = tf.io.decode_csv(value, record_defaults=record_defaults)
+        all_columns = collections.OrderedDict(zip(column_headers, columns))
+        labels = []
+        for i in range(0, len(LABEL_COLUMNS)):
+            labels.append(all_columns.pop(LABEL_COLUMNS[i]))
+        label = tf.stack(labels, axis=1)
+        features = all_columns
+        return features, label
+
+    '''Work Queue Feature'''
+    if args.workqueue and not args.tf:
+        from tensorflow.python.ops.work_queue import WorkQueue
+        work_queue = WorkQueue([filename])
+        # For multiple files：
+        # work_queue = WorkQueue([filename, filename1,filename2,filename3])
+        files = work_queue.input_dataset()
+    else:
+        files = filename
+    # Extract lines from input files using the Dataset API.
+    dataset = tf.data.TextLineDataset(files)
+    dataset = dataset.shuffle(buffer_size=20000,
+                              seed=args.seed)  # set seed for reproducing
+    dataset = dataset.repeat(num_epochs)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(parse_csv, num_parallel_calls=28)
+    dataset = dataset.prefetch(1)
```

Comment:
prefetch(2), keep aligned with other models in modelzoo.

Reply:
Changed